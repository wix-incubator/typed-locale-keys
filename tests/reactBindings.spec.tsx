import fs from 'fs';

import { render } from '@testing-library/react';
import React from 'react';

import {
  LocaleKeysProvider,
  useLocaleKeys
} from './__generated__/pregenerated/react/localeKeys';
import { Driver } from './driver';

const namespace = 'react';
let importResult: Record<string, unknown>;

let driver: Driver;

beforeAll(async () => {
  driver = new Driver();
  driver.given.namespace('react');

  await driver.when.runsCodegenCommand({
    reactHook: true
  });

  importResult = await driver.get.generatedResults<{
    common: {
      create(): string;
    };
    model: { user: { id(): string } };
  }>();
});

test('ts file is not generated', () => {
  expect(fs.existsSync(`tests/__generated__/${namespace}/localeKeys.ts`)).toBe(
    false
  );
});

test('hook and provider should be defined in file', () => {
  expect(importResult.useLocaleKeys).not.toBeUndefined();
  expect(importResult.LocaleKeysProvider).not.toBeUndefined();
});

test('provider and hook should call translation function and return its result', () => {
  const testId = 'some-translated-text';

  const SomeText = () => {
    const localeKeys = useLocaleKeys();

    return (
      <div data-testid={testId}>
        {localeKeys.common.loggedIn.message({ username: 'Bruce' })}
      </div>
    );
  };

  const renderResult = render(
    <LocaleKeysProvider tFn={driver.get.defaultTranslationFn()}>
      <SomeText />
    </LocaleKeysProvider>
  );

  expect(renderResult.getByTestId(testId).innerHTML).toContain(
    driver.get.expectedTranslationOf('common.loggedIn.message', {
      username: 'Bruce'
    })
  );
});
