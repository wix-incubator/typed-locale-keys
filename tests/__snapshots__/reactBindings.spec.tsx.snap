// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`react file (snapshot) 1`] = `
"/* eslint-disable */
/* tslint:disable */
import React from 'react';

export type ILocaleKeys = {
    common: {
      loggedIn: {
        message: (data: Record<'username', unknown>) => string, /* Hey, {{username}}, you have successfully logged in! */
      },
    },
    readingWarning: (data: Record<'reader' | 'writer', unknown>) => string, /* {{reader}} reads message from {{writer}} */
  };
const createProxyImpl = <R extends string>(
  t = (...[k]: unknown[]) => k as R,
  prevKeys = ''
): unknown =>
  new Proxy((...args: unknown[]) => t(prevKeys, ...args), {
    get: (_, key: string): unknown => {
      let nextKey = prevKeys;

      if (key !== '$value') {
        nextKey = prevKeys ? [prevKeys, key].join('.') : key;
      }

      return createProxyImpl(t, nextKey);
    },
  });

export function LocaleKeys<R extends string>(t: (...args: unknown[]) => R) {
  return createProxyImpl(t) as ILocaleKeys;
}

const LocaleKeysContext = React.createContext({} as ILocaleKeys);
export const LocaleKeysProvider: React.FC<{ translateFn?: (...args: unknown[]) => string; localeKeys?: ILocaleKeys; children?: React.ReactNode }> = ({ translateFn, localeKeys, children }) => {
    if (!translateFn && !localeKeys) { throw new Error('Either translateFn or localeKeys must be provided') }
    const value = (typeof translateFn === 'function' ? LocaleKeys(translateFn) : localeKeys) as ILocaleKeys
    return <LocaleKeysContext.Provider value={value}>{children}</LocaleKeysContext.Provider>;
  };
export const useLocaleKeys = () => React.useContext(LocaleKeysContext);
"
`;
